otherlibs/systhreads/thread.ml

```diff
@@ -19,6 +19,126 @@
+
+  let rec add_parent_key ki =
+    let l = Atomic.get parent_keys in
+    if not (Atomic.compare_and_set parent_keys l (ki :: l))
+    then add_parent_key ki
+
+  let new_key ?split_from_parent init_orphan =
+    let idx = Atomic.fetch_and_add key_counter 1 in
+    let k = idx, { Modes.Portable.portable = init_orphan } in
+    begin match split_from_parent with
+    | None -> ()
+    | Some split -> add_parent_key (KI (k, split))
+    end;
+    k
+
+  (* If necessary, grow the current domain's local state array such that [idx]
+   * is a valid index in the array. *)
+  let maybe_grow idx =
+    let state = get_tls_state () in
+    let size = Array.length state in
+    if idx < size then state
+    else begin
+      let rec compute_new_size s =
+        if idx < s then s else compute_new_size (2 * s)
+      in
+      let new_size = compute_new_size size in
+      let new_state =
+        Array.make new_size (Obj.magic_uncontended Obj_opt.none)
+      in
+      Array.blit state 0 new_state 0 size;
+      set_tls_state new_state;
+      new_state
+    end
+
+  let set (type a) (idx, _init) (x : a) =
+    (* Assures [idx] is in range. *)
+    let state = maybe_grow idx in
+    (* [Sys.opaque_identity] ensures that flambda does not look at the type of
+     * [x], which may be a [float] and conclude that the [st] is a float array.
+     * We do not want OCaml's float array optimisation kicking in here. *)
+    Array.unsafe_set state idx (Obj_opt.some (Sys.opaque_identity x))
+
+  let get (type a) ((idx, init) : a key) : a =
+    (* Assures [idx] is in range. *)
+    let state = maybe_grow idx in
+    let obj = Array.unsafe_get state idx in
+    if Obj_opt.is_some obj
+    then (Obj_opt.unsafe_get obj : a)
+    else begin
+      let v : a = init.Modes.Portable.portable () in
+      let new_obj = Obj_opt.some (Sys.opaque_identity v) in
+      let state = get_tls_state () in
+      Array.unsafe_set state idx new_obj;
+      v
+    end
+
+  type key_value : value mod portable contended =
+      KV : 'a key * (unit -> 'a) @@ portable -> key_value
+  [@@unsafe_allow_any_mode_crossing "CR with-kinds"]
+
+  let get_initial_keys () : key_value list =
+    List.map
+      (* [v] is applied exactly once in [set_initial_keys] *)
+      (fun (KI (k, split)) ->
+        let v = Obj.magic_many (split (get k)) |> Obj.magic_portable in
+        KV (k, v))
+      (Atomic.get parent_keys : key_initializer_list)
+
+  let set_initial_keys (l : key_value list) =
+    List.iter (fun (KV (k, v)) -> set k (v ())) l
+
+  (* Freshly spawned domains have uninitialized TLS state, so it would be
+     unsafe for the split functions of DLS keys registered before now to
+     use TLS. No such keys exist because we are defining the TLS module. *)
+  let _split_tls_state_on_spawn =
+    Domain.Safe.DLS.new_key ~split_from_parent:(fun () ->
+      let tls_keys = get_initial_keys () in
+      (fun () -> init (); set_initial_keys tls_keys))
+      (fun () -> ())
```

Comment:
Hmm... Creating a new DLS key (and effectively an unused DLS slot) for initializing TLS feels a bit wasteful.  It is probably not extremely expensive, but I would expect it also would not require many lines of code to add support for some sort of "do on spawn callbacks" to domain.  Perhaps the `get_initial_keys + set_initial_keys` dance could be generalized a bit to not require setting a DLS/TLS entry.  Perhaps something like `get_initializers + run_initializers` where `get_initializers` returns a list of closures to run in the new domain/thread.